File: scripts/query-unpaid-list.py
#!/usr/bin/env python3
"""Standalone unpaid list query against erp.sqlite."""
import json
import sqlite3
import sys
from datetime import date

DB_PATH = sys.argv[3] if len(sys.argv) > 3 else "/home/ubuntu/teammillimeter-erp/data/erp.sqlite"
START = sys.argv[1] if len(sys.argv) > 1 else f"{date.today().year}-{date.today().month:02d}-01"
END = sys.argv[2] if len(sys.argv) > 2 else None

if END is None:
    y, m = map(int, START.split("-")[:2])
    if m == 12:
        END = f"{y}-12-31"
    else:
        from calendar import monthrange
        END = f"{y}-{m:02d}-{monthrange(y, m)[1]:02d}"


def get_unpaid(sale):
    amount = float(sale.get("amount") or 0)
    paid = float(sale.get("paid") if sale.get("paid") is not None else sale.get("basePaid") or 0)
    return max(amount - paid, 0)


def fmt_krw(n):
    return f"{int(round(n)):,}"


conn = sqlite3.connect(DB_PATH)
row = conn.execute("SELECT payload FROM erp_domain_state WHERE domain = 'sales'").fetchone()
conn.close()

if not row:
    print("No sales domain payload found")
    sys.exit(1)

payload = json.loads(row[0])
sales = payload.get("sales") or []

rows = []
for sale in sales:
    unpaid = get_unpaid(sale)
    if unpaid <= 0:
        continue
    d = str(sale.get("date") or "")[:10]
    if d < START or d > END:
        continue
    rows.append({
        "date": d,
        "client": str(sale.get("client") or ""),
        "site": str(sale.get("site") or ""),
        "unpaid": unpaid,
    })

rows.sort(key=lambda r: (-r["unpaid"], r["date"]), reverse=False)
rows.sort(key=lambda r: (-r["unpaid"], r["date"]))

total = sum(r["unpaid"] for r in rows)
print(f"DB: {DB_PATH}")
print(f"Period: {START}~{END}")
print(f"Total unpaid: {fmt_krw(total)}? ({len(rows)}?)")
for r in rows[:20]:
    site = f" / {r['site']}" if r["site"] else ""
    print(f"- {r['client']}{site} ({r['date']}): {fmt_krw(r['unpaid'])}?")
if len(rows) > 20:
    print(f"… ? {len(rows) - 20}?")

File: scripts/query-unpaid-list.mjs
#!/usr/bin/env node
/**
 * Standalone unpaid list query against erp.sqlite (no erpChatTools import).
 * Usage: node scripts/query-unpaid-list.mjs [startDate] [endDate]
 * Default: current month (KST-ish via server local date).
 */
import fs from "fs";
import path from "path";
import { createRequire } from "module";

const require = createRequire(import.meta.url);
const Database = require("better-sqlite3");

function todayISO() {
  const d = new Date();
  return `${d.getFullYear()}-${String(d.getMonth() + 1).padStart(2, "0")}-${String(d.getDate()).padStart(2, "0")}`;
}

function monthRangeISO(offset = 0) {
  const today = todayISO();
  const match = /^(\d{4})-(\d{2})-(\d{2})$/.exec(today);
  if (!match) return { startDate: today, endDate: today };
  const year = Number(match[1]);
  const month = Number(match[2]) - 1;
  const date = new Date(year, month + offset, 1);
  const startDate = `${date.getFullYear()}-${String(date.getMonth() + 1).padStart(2, "0")}-01`;
  const endDateObj = new Date(date.getFullYear(), date.getMonth() + 1, 0);
  const endDate = `${endDateObj.getFullYear()}-${String(endDateObj.getMonth() + 1).padStart(2, "0")}-${String(endDateObj.getDate()).padStart(2, "0")}`;
  return { startDate, endDate };
}

function formatKRW(n) {
  return Number(n || 0).toLocaleString("ko-KR");
}

function getUnpaid(sale) {
  const amount = Number(sale.amount) || 0;
  const paid = Number(sale.paid ?? sale.basePaid ?? 0) || 0;
  return Math.max(amount - paid, 0);
}

function resolveDbPath() {
  const candidates = [
    process.env.ERP_SQLITE_PATH,
    path.resolve("data/erp.sqlite"),
    path.resolve("../data/erp.sqlite"),
    "/home/ubuntu/teammillimeter-erp/data/erp.sqlite",
  ].filter(Boolean);
  for (const p of candidates) {
    if (fs.existsSync(p)) return p;
  }
  throw new Error(`erp.sqlite not found. Tried: ${candidates.join(", ")}`);
}

function loadSales(db) {
  const row = db.prepare("SELECT payload FROM erp_domain_state WHERE domain = 'sales'").get();
  if (!row?.payload) return [];
  try {
    const parsed = JSON.parse(row.payload);
    return Array.isArray(parsed.sales) ? parsed.sales : Array.isArray(parsed) ? parsed : [];
  } catch {
    return [];
  }
}

const args = process.argv.slice(2);
let startDate = args[0];
let endDate = args[1];
if (!startDate || !endDate) {
  const range = monthRangeISO(0);
  startDate = range.startDate;
  endDate = range.endDate;
}

const dbPath = resolveDbPath();
const db = new Database(dbPath, { readonly: true });
const sales = loadSales(db);

const unpaidRows = sales
  .map((sale) => ({
    id: sale.id,
    date: String(sale.date || "").slice(0, 10),
    client: String(sale.client || ""),
    site: String(sale.site || ""),
    amount: Number(sale.amount) || 0,
    paid: Number(sale.paid ?? sale.basePaid ?? 0) || 0,
    unpaid: getUnpaid(sale),
  }))
  .filter((row) => {
    if (row.unpaid <= 0) return false;
    if (startDate && row.date < startDate) return false;
    if (endDate && row.date > endDate) return false;
    return true;
  })
  .sort((a, b) => b.unpaid - a.unpaid || String(b.date).localeCompare(String(a.date)));

const totalUnpaid = unpaidRows.reduce((sum, row) => sum + row.unpaid, 0);
const maxRows = 20;

console.log(`DB: ${dbPath}`);
console.log(`Period: ${startDate}~${endDate}`);
console.log(`Total unpaid: ${formatKRW(totalUnpaid)}? (${unpaidRows.length}?)`);
for (const row of unpaidRows.slice(0, maxRows)) {
  const site = row.site ? ` / ${row.site}` : "";
  console.log(`- ${row.client}${site} (${row.date}): ${formatKRW(row.unpaid)}?`);
}
if (unpaidRows.length > maxRows) {
  console.log(`… ? ${unpaidRows.length - maxRows}?`);
}

db.close();
